enhance(ECDH): reconstruct ECDH

AdoptOSS opened the pull request:
## Metadate
ABI Change: No
Supersed #1928

## Changelog
### Common
common: 移动特定于 QQ 平台的实现到 OicqECDH，重新设计 ECDH，使 ECDH 类只针对算法本身，而不过多包括 QQ 协议的使用细节

感觉这样结构似乎更清晰一点

### JVM-like Platform
jvm: 尝试优先使用平台加密实现，可能改善性能（and fallback to `BouncyCastleProvider` explicitly）
jvm & android: 使用 curveName `secp256r1` 代替 `prime256v1`，前者在 Java 中更常用，可以被更多的 JCE 实现所识别（虽然都是指同一条曲线）
android: 使用系统自带的实现以减少依赖，并尝试兼容 Android P+ 版本
> Starting in Android P, we plan to deprecate some functionality from the BC provider that's duplicated by the AndroidOpenSSL (also known as Conscrypt) provider.
> ([View origin](https://android-developers.googleblog.com/2018/03/cryptography-changes-in-android-p.html))

所以 Android P+ 上也是可以使用自带的加密库的，只是不应该指定 `provider = "BC"`

### Native
native(openssl): 中间储存密钥时，保留 OpenSSL 内部数据结构，而不反复 new & free，以便提高性能
    （为了实现智能指针，需要用到 `@ExperimentalStdlibApi` 的 `createCleaner`，但个人认为这种风险应该可以接受）
native(openssl): 直接使用 point/bignum 到 bytes 的转换，避免了使用 hex string 作为中间层，提高效率（`EC_POINT_point2oct` / `EC_POINT_oct2point`）
~mingwx64: 新增使用系统自带实现BCrypt~ （没有启用，写到快结束才发现`BCRYPT_KDF_RAW_SECRET`要Win8.1及以上版本才能用，而Oicq用的又是非常见的密钥派生方式，似乎必须要拿到 raw secret 手动处理才能可用 😢，还没想好通过何种方式提供 option，先放这里了）

AdoptOSS committed:
native: 优化 build script 及 actions，修复 ECDH 相关资源释放问题 (#2110)

* build: fix build on Windows

* test: fix ContentEqualsTest on native

* build: allow disabling targets with property

* fix: free ECDH-related resources properly on native
avoid memory leaking (usually on *nix) or crash (usually on Windows)

Signed-off-by: AdoptOSS <adoptoss@outlook.com>

* build(workflow): remove invalid options

* fix(styling): `mirai.target` property

* build(workflow): try to limit memory usage during mirai-console:tools:gradle-plugin:integTest

* enhance(test): use buildList

* build(workflow): retry

Co-authored-by: ArcticLampyrid <arcticlampyrid@outlook.com>

Him188 committed:
fix build

Him188 committed:
Minor docs update for NetworkHandler

AdoptOSS committed:
fix: platform socket on Windows (#2121)

* fix: platform socket on Windows

* fix(workflow): use win platform ssl provider instead of openssl for cURL
openssl can't use system's native CA store by default

Commit:
enhance(ECDH): reconstruct ECDH
common: 移动特定于 QQ 平台的实现到 OicqECDH，重新设计 ECDH，使 ECDH 类只针对算法本身，而不过多包括 QQ 协议的使用细节
jvm: 尝试优先使用平台加密实现，可能改善性能
jvm & android: 使用 curveName `secp256r1` 代替 `prime256v1`，前者在 Java 中更常用，可以被更多的 JCE 实现所识别（虽然都是指同一条曲线）
android: 使用系统自带的实现以减少依赖，并尝试兼容 Android P+ 版本
native: 中间储存时保留OpenSSL内部结构而不反复 new & free，提高性能
    （为了实现智能指针，需要用到 `@ExperimentalStdlibApi` 的 `createCleaner`，但这种风险应该可以接受）
native: 直接使用 point/bignum 到 bytes 的转换，避免了 hex string 作为中间层，提高效率

Commit:
test(ECDH): fix AndroidTest

Commit:
enhance(ECDH): add bcrypt impl for ECDH, not enabled
As BCRYPT_KDF_RAW_SECRET (essential feature in our case) is available from Windows 8.1 and onwards, it's not suitable to enable it by default now.

Him188 reviewed (commented):
这个改动很重要，需要增加相关测试（确保协议上能跑）

AdoptOSS commented:
实际登录测试

#### Windows
Linux/MacOS下应该也不会有问题
- [x] jvm:
-  -  [x] JceECDH (Default Provider, 我这里是 Sun Security Provider)
-  -  [x] JceECDH (BC Provider)
- [x] native:
-  -  [x] OpenSslECDH
-  -  [x] BcryptECDH (Win10下登录测试成功，虽然最后没启用)

#### Android
待测
- [ ] APILevel < 23, Before Android P (System BC Provider)
- [ ] APILevel >=23, Android P and onward (Default Provider)

#### Code
```kotlin
fun main(): Unit = runBlocking {
    initMirai() // native only
    val bot = newBot(QQ_ACCOUNT, QQ_PASSWORD) {
        parentCoroutineContext = coroutineContext
        fileBasedDeviceInfo(DEVICE_INFO_PATH)
        disableContactCache()
    }
    bot.login()

    getchar(); // native only

    /* jvm only
    withContext(Dispatchers.IO) {
        System.`in`.read()
    }
    */

    bot.closeAndJoin()
}
```

sandtechnology commented on mirai-core/src/commonMain/kotlin/network/components/EcdhInitialPublicKeyUpdater.kt:
```diff
@@ -67,16 +66,12 @@ internal class EcdhInitialPublicKeyUpdaterImpl(
+    var oicqEcdh: OicqECDH? = null
+    override fun getOicqECDH(): OicqECDH {
+        if (oicqEcdh == null) {
             error("Calling getECDHWithPublicKey without calling refreshInitialPublicKeyAndApplyECDH")
```
`Calling getECDHWithPublicKey without calling refreshInitialPublicKeyAndApplyECDH` 应该更改？

AdoptOSS replied:
错误信息的确要修改
`oicq`还是`QQ`的问题，考虑到Mirai本身没有相关的 guide，且其他代码也很大程度混用二者，可能需要项目组来最终决定

Him188 replied:
我觉得用 QQ, 因为目前有 QQAndroidBot 这个命名

sandtechnology reviewed (changes requested):
此外还有有个命名上的问题 使用 `oicq` 还是 `QQ`？因为在变动中看见了两种用法

AdoptOSS commented:
但这样就会变成 QQECDH，亦或者写作 QQEcdh 来明确单词边界？

Him188 commented:
按  Kotlin命名规范，两个字母的词全部大写，三个字母及以上首字母大写

Commit:
style(Ecdh): obey official Kotlin coding conventions
> When using an acronym as part of a declaration name, capitalize it if it consists of two letters (IOStream); capitalize only the first letter if it is longer (XmlFormatter, HttpInputStream).
> [View origin](https://kotlinlang.org/docs/coding-conventions.html)

Him188 commented:
native 得 rebase onto dev, 你先继续基于之前的 branch 写吧, 写完了我单独 squash 合并

AdoptOSS commented:
大概就这样了，打算暂且只重构一下结构，提升一下 Android 和 JVM Desktop下的体验
BCrypt实现可以不放里面，毕竟 native 下类似可选择 engine 的设计可能会很麻烦（看起来很多链接过程是static的）

AdoptOSS commented:
似乎在 MacOS 上也可以使用 [Apple CryptoKit](https://developer.apple.com/documentation/cryptokit) 的 [P256](https://developer.apple.com/documentation/cryptokit/p256) 实现
不过不是 Apple 用户，就不写了（没法调试）

Requirements:
- iOS 13.0+
- iPadOS 13.0+
- macOS 10.15+
- Mac Catalyst 13.0+
- tvOS 13.0+
- watchOS 6.0+
- Xcode 11.0+
